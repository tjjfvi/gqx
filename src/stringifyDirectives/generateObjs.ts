
import { Obj } from ".";
import { Context } from "..";

export default (ctx: Context) =>
  Object.entries(ctx.objectTypes).map(([type, fields]): Obj => ({
    type,
    props: fields.map(f => ({
      type,
      prop: f.name.value,
      dirs: (f.directives || []).map(d => ({
        type,
        prop: f.name.value,
        dirName: d.name.value,
        dirArgs: d.arguments?.slice(),
      }))
    })),
  }))
