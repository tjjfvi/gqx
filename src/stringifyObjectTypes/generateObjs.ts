
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Id[], Context] => {
  const ids: Id[] = [];
  const objs: Obj[] =
    Object.keys(ctx.objectTypes)
      .sort()
      .map(name => {
        const { fields, unions, implements: impls } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          ids.push(id);
          const [{ name: { value: typeName } }, wrap,, wrapHKT] = unwrapType(field.type);
          const shallow = !(typeName in ctx.objectTypes);
          let prop: Prop = {
            id,
            type: typeName,
            shallow,
            wrap,
            wrapHKT,
            args: field.arguments?.slice() ?? [],
            directives: field.directives?.slice() ?? [],
            loc: field.loc,
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const shallowProps = props.filter(p => p.shallow);
        const deepProps = props.filter(p => !p.shallow);
        const obj: Obj = {
          type: name,
          shallowProps,
          deepProps,
          props,
          unions,
          implements: impls,
          isBase: ctx.baseTypes.includes(name),
        };
        return obj;
      });
  return [objs, ids, ctx];
}
