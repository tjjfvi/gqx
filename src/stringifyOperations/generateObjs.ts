
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Context] => {
  const objs: Obj[] =
    ctx.operations
      .map(([prop, name]) => {
        const { fields } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          const [{ name: { value: typeName } }, wrap] = unwrapType(field.type);
          let prop: Prop = {
            id,
            type: typeName,
            wrap,
            args: field.arguments?.slice() ?? [],
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const obj: Obj = {
          type: name,
          prop,
          props,
        };
        return obj;
      })
  return [objs, ctx];
}
