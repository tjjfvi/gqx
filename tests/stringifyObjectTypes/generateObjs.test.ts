
import generateObjs from "../../src/stringifyObjectTypes/generateObjs";

test("", () => {
  const ctx = utils.getOperations(["extensive"]);
  const [objs, ids] = generateObjs(ctx);
  expect(ids).toMatchSnapshot();
  expect(objs.map(o => ({ n: o.type, s: o.shallowProps.length, d: o.deepProps.length }))).toMatchSnapshot();
  expect(objs).toMatchSnapshot();
})
