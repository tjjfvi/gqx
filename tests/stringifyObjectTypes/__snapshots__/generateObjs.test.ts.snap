// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  Object {
    "prop": "node",
    "type": "BinaryTree",
  },
  Object {
    "prop": "left",
    "type": "BinaryTree",
  },
  Object {
    "prop": "right",
    "type": "BinaryTree",
  },
  Object {
    "prop": "a",
    "type": "DeepOnly",
  },
  Object {
    "prop": "b",
    "type": "DeepOnly",
  },
  Object {
    "prop": "id",
    "type": "Recursive",
  },
  Object {
    "prop": "next",
    "type": "Recursive",
  },
  Object {
    "prop": "a",
    "type": "ShallowOnly",
  },
  Object {
    "prop": "b",
    "type": "ShallowOnly",
  },
]
`;

exports[` 2`] = `
Array [
  Object {
    "d": 2,
    "n": "BinaryTree",
    "s": 1,
  },
  Object {
    "d": 2,
    "n": "DeepOnly",
    "s": 0,
  },
  Object {
    "d": 1,
    "n": "Recursive",
    "s": 1,
  },
  Object {
    "d": 0,
    "n": "ShallowOnly",
    "s": 2,
  },
]
`;

exports[` 3`] = `
Array [
  Object {
    "deepProps": Array [
      Object {
        "id": Object {
          "prop": "left",
          "type": "BinaryTree",
        },
        "shallow": false,
        "type": "BinaryTree",
        "wrap": [Function],
      },
      Object {
        "id": Object {
          "prop": "right",
          "type": "BinaryTree",
        },
        "shallow": false,
        "type": "BinaryTree",
        "wrap": [Function],
      },
    ],
    "shallowProps": Array [
      Object {
        "id": Object {
          "prop": "node",
          "type": "BinaryTree",
        },
        "shallow": true,
        "type": "String",
        "wrap": [Function],
      },
    ],
    "type": "BinaryTree",
  },
  Object {
    "deepProps": Array [
      Object {
        "id": Object {
          "prop": "a",
          "type": "DeepOnly",
        },
        "shallow": false,
        "type": "ShallowOnly",
        "wrap": [Function],
      },
      Object {
        "id": Object {
          "prop": "b",
          "type": "DeepOnly",
        },
        "shallow": false,
        "type": "ShallowOnly",
        "wrap": [Function],
      },
    ],
    "shallowProps": Array [],
    "type": "DeepOnly",
  },
  Object {
    "deepProps": Array [
      Object {
        "id": Object {
          "prop": "next",
          "type": "Recursive",
        },
        "shallow": false,
        "type": "Recursive",
        "wrap": [Function],
      },
    ],
    "shallowProps": Array [
      Object {
        "id": Object {
          "prop": "id",
          "type": "Recursive",
        },
        "shallow": true,
        "type": "ID",
        "wrap": [Function],
      },
    ],
    "type": "Recursive",
  },
  Object {
    "deepProps": Array [],
    "shallowProps": Array [
      Object {
        "id": Object {
          "prop": "a",
          "type": "ShallowOnly",
        },
        "shallow": true,
        "type": "String",
        "wrap": [Function],
      },
      Object {
        "id": Object {
          "prop": "b",
          "type": "ShallowOnly",
        },
        "shallow": true,
        "type": "Number",
        "wrap": [Function],
      },
    ],
    "type": "ShallowOnly",
  },
]
`;
